TASK: Amends mautic:maintenance:cleanup help text

putzwasser opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 5.x)
* a.b for any bug fixes (e.g. 4.4, 5.1)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [X]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/mautic-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

#### Description:

This PR amends the `mautic:maintenance:cleanup` help text. The text explains now which data get purged. It also explains the effects of --gdpr in a clear way.

#### Steps to test this PR:

<!--
This part is really important. If you want your PR to be merged, take the time to write very clear, annotated and step by step test instructions. Do not assume any previous knowledge - testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Run `php bin/console help mautic:maintenance:cleanup`

<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->

putzwasser committed:
TASK: Amends mautic:maintenance:cleanup help text

putzwasser committed:
TASK: Adds <info> tag to event identifier

RCheesley commented on app/bundles/CoreBundle/Command/CleanupMaintenanceCommand.php:
```diff
@@ -46,17 +46,32 @@ protected function configure()
             ->setHelp(
                 <<<'EOT'
-                The <info>%command.name%</info> command dispatches the CoreEvents::MAINTENANCE_CLEANUP_DATA event in order to purge old data (data must be supported by event listeners as not all data is applicable to be purged).
+<info>%command.name%</info> purges records of anonymous contacts (<comment>unless <info>--gdpr</info> flag is set</comment>) that are older than 365 days.
```
```suggestion
<info>%command.name%</info> purges records of anonymous contacts (<comment>unless the <info>--gdpr</info> flag is set</comment>) that are older than 365 days.
```

RCheesley commented on app/bundles/CoreBundle/Command/CleanupMaintenanceCommand.php:
```diff
@@ -46,17 +46,32 @@ protected function configure()
+Shows you how many records of anonymous contacts <info>%command.name%</info> will purge.
 
-<info>php %command.full_name% --days-old=365 --dry-run</info>
+The <info>%command.name%</info> command dispatches the <info>CoreEvents::MAINTENANCE_CLEANUP_DATA</info> event in order to purge old data (data must be supported by event listeners as not all data is applicable to be purged).
```
```suggestion
The <info>%command.name%</info> command dispatches the <info>CoreEvents::MAINTENANCE_CLEANUP_DATA</info> event in order to purge old data (data must be supported by event listeners, as not all data is applicable to be purged).
```

RCheesley reviewed (changes requested):
Hi there! Thanks for the PR, I've made a couple of small suggestions for grammar fixes.

Could you also fix the typo on line 44:

fullfil -> fulfill (assuming US spelling)

putzwasser committed:
Fix: Grammer

Co-authored-by: Ruth Cheesley <ruth@ruthcheesley.co.uk>

putzwasser committed:
FIX: Grammar

Co-authored-by: Ruth Cheesley <ruth@ruthcheesley.co.uk>

putzwasser committed:
TASK: Amend flag help text.

putzwasser committed:
FEAT: Makes --gdpr purge threshold configurable

putzwasser commented:
@RCheesley: Done.

I also removed the hard coded day threshold for `--gdpr`.
Created `mautic.gdpr_user_purge_threshold` config parameter with a default value of 1095 and used it in the command.

This makes
- the threshold adjustable to specific needs and
- keeps the command safe (i.e. by not allowing to set a day threshold along with `--gdpr`).

I don't know if I made the changes the Mautic/Symfony way. Maybe @mabumusa1 or @escopecz could take a look.

RCheesley reviewed (dismissed):
Thanks for making the changes requested - I'll leave it to @escopecz and @mabumusa1 to provide feedback on the other aspects. Please also check the failing tests @putzwasser 🚀

escopecz commented on app/bundles/CoreBundle/Command/CleanupMaintenanceCommand.php:
```diff
@@ -25,6 +25,7 @@ public function __construct(
         PathsHelper $pathsHelper,
         CoreParametersHelper $coreParametersHelper
     ) {
+    	$this->coreParametersHelper = $coreParametersHelper;
```
Just add `private` keyword to the line 26 and remove this line to avoid this static analysis error.

putzwasser replied:
Done.

putzwasser committed:
FIX: static analysis error

escopecz reviewed (dismissed):
The code changes look great to me. Thank you! 👍

kuzmany reviewed (changes requested):
I'm fine with changing the text strings, but we have already submitted a different solution for using the "days-old" parameter for GDPR. You can find it on GitHub at this link: https://github.com/mautic/mautic/pull/12332/

putzwasser commented:
> but we have already submitted a different solution for using the "days-old" parameter for GDPR

I see the danger that someone could unintentionally delete lead data.

A hard coded value of 1095 days seemed to me to be some sort of safety net. To make the strict limit more flexible, I moved the value to the configuration file.

This makes it possible to adjust the limit without `bin/console mautic:maintenance:cleanup --gdpr --days-old=36` (hitting enter before typing `5`) deleting almost all users. There are many possible mistakes one could make here—all of them come with a severe loss of data.

Adjusting the setting is a tiny bit more cumbersome than using `--days-old`, but you only need to set this value once. From a GDPR compliance point of view there is no need to have multiple/different day thresholds.

kuzmany commented:
@putzwasser I understand and accept your explanation as well. It's worth noting that making a typo in a local parameter can also lead to the same outcome.

@escopecz Can you determine which solution is better for you: this or https://github.com/mautic/mautic/pull/12332? I am referring to the ability to change the timeframe for deleting contacts with the `--gdpr` flag.  Our solution utilizes the "--days-old" parameter in the command. It is an option that, if not used, will default to a duration of three years, just as it was before without any backward compatibility changes.

escopecz commented:
Why can't we have both options? The config value as the default and the command param that can override the default? 

@putzwasser do you think you could cherry-pick @kuzmany 's commit from https://github.com/mautic/mautic/pull/12332 and implement that command param to this PR? Then we can merge this PR, close that one and be all happy 🙂

kuzmany commented:
@escopecz I don't think we need both, in my opinion. We can shift our focus to something else while still achieving the primary objective through either the first or second option.

kuzmany commented:
@putzwasser Are you able to finish the PR to a mergeable status (unit tests, etc.)?
Then we can close https://github.com/mautic/mautic/pull/12332 and I would like to use this solution.

putzwasser commented:
> Are you able to finish the PR to a mergeable status (unit tests, etc.)?

Kinda. I'm a bit short of time, currently. Plus, I've yet to write my first unit test 🙈

So, support would be greatly appreciated.

What else is needed? I.e., what do you mean by `etc`? :)

kuzmany committed:
Add test for GDPR maintenance command in CleanupMaintenanceCommandTest

kuzmany committed:
Merge branch 'refs/heads/5.x' into fork/enhance-maintenance-help-text

# Conflicts:
#	app/bundles/CoreBundle/Command/CleanupMaintenanceCommand.php

kuzmany commented:
@putzwasser @escopecz I've added unit tests and closed duplicated PRs https://github.com/mautic/mautic/pull/12332. Should be ready to merge

RCheesley commented on app/bundles/CoreBundle/Command/CleanupMaintenanceCommand.php:
```diff
@@ -46,23 +46,38 @@ protected function configure(): void
                     ),
-                    new InputOption('dry-run', 'r', InputOption::VALUE_NONE, 'Do a dry run without actually deleting anything.'),
-                    new InputOption('gdpr', 'g', InputOption::VALUE_NONE, 'Delete data to fullfil GDPR requirement.'),
+                    new InputOption('dry-run', 'r', InputOption::VALUE_NONE, 'Perfroms a dry run. Shows no. of affected rows. Won\'t actually delete anything.'),
```
```suggestion
                    new InputOption('dry-run', 'r', InputOption::VALUE_NONE, 'Performs a dry run. Shows no. of affected rows. Won\'t actually delete anything.'),
```

RCheesley commented on app/bundles/CoreBundle/Command/CleanupMaintenanceCommand.php:
```diff
@@ -46,23 +46,38 @@ protected function configure(): void
-                    new InputOption('dry-run', 'r', InputOption::VALUE_NONE, 'Do a dry run without actually deleting anything.'),
-                    new InputOption('gdpr', 'g', InputOption::VALUE_NONE, 'Delete data to fullfil GDPR requirement.'),
+                    new InputOption('dry-run', 'r', InputOption::VALUE_NONE, 'Perfroms a dry run. Shows no. of affected rows. Won\'t actually delete anything.'),
+                    new InputOption('gdpr', 'g', InputOption::VALUE_NONE, 'Deletes records of inactive users to fullfill GDPR requirements.'),
```
```suggestion
                    new InputOption('gdpr', 'g', InputOption::VALUE_NONE, 'Deletes records of inactive users to fulfill GDPR requirements.'),
```

RCheesley reviewed (changes requested):
Please fix these spelling errors (one that I flagged up previously). Thanks!

kuzmany committed:
Apply suggestions from code review

Co-authored-by: Ruth Cheesley <ruth@ruthcheesley.co.uk>

putzwasser committed:
TASK: Adds tests for keeping active leads/contacts

- Adds assertion messages
- Refactors lead creation completely into `createLead` to stick to DRY principle

putzwasser committed:
Merge branch '5.x' into enhance-maintenance-help-text

putzwasser commented:
@kuzmany  Thanks for creating the tests. I had a look into the unit tests and tried to understand how these are done, to be able to include them for any next PR.

I added two more assertions: Tests for keeping leads/contacts that are still active. I thought these tests are important, too, (and they served as a small exercise) to prevent purging leads accidentally.

RCheesley commented:
Thanks for working together on this folks! There's a conflict which needs to be resolved, then we should be able to get this reviewed and merged!

putzwasser committed:
Merge branch '5.x' into enhance-maintenance-help-text

RCheesley committed:
Merge branch '5.x' into enhance-maintenance-help-text

RCheesley reviewed (approved):
Good to go, thanks for all the work on this folks!
